share/lib/python/neuron/coreneuron.py

```diff
@@ -0,0 +1,82 @@
+
+# Properties settable by user
+enable = False     # Use CoreNEURON when calling ParallelContext.psolve(tstop).
+gpu = False        # Activate GPU computation.
+cell_permute = 1   # 0 no permutation; 1 optimize node adjacency
+                   # 2 optimize parent node adjacency (only for gpu = True)
+warp_balance = 0   # Number of warps to balance. (0 no balance)
+verbose = 0        # 0 quiet
+prcellstate = -1   # Output prcellstate information for the gid at t=0
+                   # and at tstop. -1 means no output
+
+def property_check(tstop):
+  '''
+  Check type and value in range for the user properties.
+  '''
+  global enable, gpu, cell_permute, warp_balance, verbose, prcellstate
+  tstop = float(tstop) # error if can't be a float
+  enable = bool(int(enable))
+  gpu = bool(gpu)
+  cell_permute = int(cell_permute)
+  assert(cell_permute in range(3))
+  assert(cell_permute in (range(3) if gpu else range(2)))
+  warp_balance = int(warp_balance)
+  assert(warp_balance >= 0)
+  verbose = int(verbose)
+  prcellstate = int(prcellstate)
+
+def nrncore_arg(tstop):
+  """
+  Return str that can be used for pc.nrncore_run(str)
+  based on user properties and current NEURON settings.
+  :param tstop: Simulation time (ms)
+
+  """
+  from neuron import h
+  arg = ''
+  property_check(tstop)
+  if not enable:
+    return arg
+
+  # args derived from user properties
+  arg += ' --gpu' if gpu else ''
+  arg += ' --tstop %g' % tstop
+  arg += (' --cell-permute %d' % cell_permute) if cell_permute > 0 else ''
+  arg += (' --nwarp %d' % warp_balance) if warp_balance > 0 else ''
+  arg += (' --prcellgid %d' % prcellstate) if prcellstate >= 0 else ''
+  arg += (' --verbose %d' % verbose) if verbose > 0 else ''
+
+  # args derived from current NEURON settings.
+  pc = h.ParallelContext()
+  cvode = h.CVode()
+  arg += ' --mpi' if pc.nhost() > 1 else ''
```

Comment:
whats if neuron is launched with mpi but with only one host? in that case tt's safe to run coreneuron also with mpi?

Reply:
It would be ok. but I think it is not necessary for coreneuron to run also with mpi in this case of single rank. And it is strangely difficult to figure out if MPI is active when there is only one host. Perhaps there should be a ```ParalleContext.mpi_active()``` method.